### PR TITLE
Linked list

### PR DESCRIPTION
**Pull request by carloscadena**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+from setuptools import setup
+
+dependencies = ['pytest', 'pytest-cov']
```
Comment (nhuntwalker) [inline setup.py:3]: Why are these required for installation of this package? Aren't these only used in testing?

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+extra_packages = {
+    'testing': ['tox']
+}
+
+setup(
+    name='Linked List',
```
Comment (nhuntwalker) [inline setup.py:9]: Is this an appropriate name for this package, which will house multiple data structures and even a few algorithms?

File: src/linked_list.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        else:
+            self.data = data
+
+    def push(self, val):
+        """pushes a value to top of linked list and makes new node the head"""
+        current = self.head
+        newl = LinkedList(val)
+        newl.next_node = current
+        self.head = newl
+        return self.head
```
Comment (nhuntwalker) [inline src/linked_list.py:26]: Why is this function returning anything?

**Comment by pasaunders** [comment]
Head should be a node a node, not a value, and certainly not multiple values. You have a lot of simple linter errors. You need to be able to init an empty list. Your tests all pass, but they're very narrow. Expand their scope.
How efficient is it to make every node a new instance of LinkedList?

File: src/linked_list.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            self.head = current.next_node
+            return tem_data
+        except AttributeError:
+            flag = 0
+        raise IndexError('linked list is empty')
+
+    def size(self):
+        """gets the size of the list"""
+        current = self.head
+        count = 1
```
Comment (nhuntwalker) [inline src/linked_list.py:43]: What if I have a linked list with nothing in it?

File: src/linked_list.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        current = self.head
+        while current.next_node is not None:
+            if current.data == val:
+                return current
+            else:
+                current = current.next_node
+        if current.data == val:
+            return current
+        else:
+            current = current.next_node
```
Comment (nhuntwalker) [inline src/linked_list.py:60]: Probably don't need this last `else` block. If you reach the final node and it isn't the value you want, raise the proper exception.

File: src/linked_list.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        raise Exception('Value does not exist in Linked List')
```
Comment (nhuntwalker) [inline src/linked_list.py:61]: There is a more specific exception that can be used here. The general `Exception` object isn't as useful or descriptive.

File: src/linked_list.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+    def remove(self, node):
+        """removes a given node from the linked list"""
+        current = self.head
+        previouscurrent = None
+        if node == current:
+            self.head = current.next_node
+            return self.head
```
Comment (nhuntwalker) [inline src/linked_list.py:69]: You shouldn't be returning anything with this function.

File: src/linked_list.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        while current.next_node is not None:
+            if current == node:
+                previouscurrent.next_node = current.next_node
+                return self.head
```
Comment (nhuntwalker) [inline src/linked_list.py:73]: Return nothing.

File: src/linked_list.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+            else:
+                previouscurrent = current
+                current = current.next_node
+        if node == current:
+            previouscurrent.next_node = None
+            return self.head
```
Comment (nhuntwalker) [inline src/linked_list.py:79]: Return nothing.

File: src/linked_list.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+        raise Exception('Node does not exist in Linked List')
```
Comment (nhuntwalker) [inline src/linked_list.py:81]: Choose a more specific exception.

File: src/linked_list.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+    def display(self):
+        """displays the data from linked list as a string"""
+        current = self.head
+        output = ''
+        while current.next_node is not None:
+            output += str(current.data) + ', '
+            current = current.next_node
+        output += str(current.data)
+        return output
```
Comment (nhuntwalker) [inline src/linked_list.py:91]: Output is missing parentheses in the string, as per the example given in the assignment documentation.

File: src/linked_list.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+    def __len__(self):
+        """returns the length of the linked list"""
+        length = self.size()
+        return length
```
Comment (nhuntwalker) [inline src/linked_list.py:96]: Why not just `return self.size()`?

File: src/linked_list.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+    def __repr__(self):
+        """prints the data as a string to stdout"""
+        output_list = self.head.display()
+        return output_list
```
Comment (nhuntwalker) [inline src/linked_list.py:101]: Same thing here, why not just `return self.head.display()`
